### ISSUE TITLE
bug: Invisible name problem

### ISSUE THREAD

**Issue by I2rys** [revoltchat/revite#666]
There is a problem with invisible name If you change your name to Invisible characters by pressing space bar 2x/4x then save it. No one will be able to send a friend request to you as well as accept your friend request. It will always return 404

**Comment by LedaThemis** [comment]
Tried it with a newly created account and it worked
```json
{
   "_id":"01G4WZVH9H6AJXG0AJPQTJ0GJA",
   "username":"    ",
   "badges":0,
   "relationship":"Friend"
}
```
This was the response that I got after accepting the friend request.

what @I2rys is describing seems to occur only when changing username to spaces
```json
{
  "error": {
    "code": 404,
    "reason": "Not Found",
    "description": "The requested resource could not be found."
  }
}
```
This was the response I got after trying to accept a friend request from an account with its username changed to spaces.

```json
{
   "_id":"01G4WZVH9H6AJXG0AJPQTJ0GJA",
   "username":"    ",
   "badges":0,
   "relationship":"Friend",
   "online":true
}
```
This is the user object of the first account (empty name at account creation)

```json
{
   "_id":"01G4WY1RKHXAW8QBFCGA16PWJS",
   "username":"",
   "badges":0,
   "relationship":"Incoming",
   "online":true
}
```
This is the user object of the second account (empty name after username change)

my guess is that the issue lies with the empty username(`""`), which affects the following code
https://github.com/revoltchat/revolt.js/blob/1b592ef8127a0f97310d4085d50db829f7ab05fd/src/maps/Users.ts#L132-L136
```js
    async addFriend() {
        return await this.client.api.put(
            `/users/${this.username as ""}/friend`,
        );
    }
```

and because the username is an empty string, it causes the API endpoint to be invalid
`PUT https://api.revolt.chat/users//friend`

Which in turn results in the following response
```json
{
   "error":{
      "code":404,
      "reason":"Not Found",
      "description":"The requested resource could not be found."
   }
}
```
~~I am not sure why several spaces become an empty string while changing username, while keeping the username intact in account creation~~ (check below), the only difference is the API methods used

https://github.com/revoltchat/revolt.js/blob/1b592ef8127a0f97310d4085d50db829f7ab05fd/src/Client.ts#L301-L306
```js
    async completeOnboarding(data: DataOnboard, loginAfterSuccess?: boolean) {
        await this.api.post("/onboard/complete", data);
        if (loginAfterSuccess) {
            await this.websocket.connect();
        }
    }
```
in account creation `POST` is used
https://github.com/revoltchat/revite/blob/12b97160433151c188b077bd3e85fc2ee35bbd99/src/context/intermediate/popovers/ModifyAccount.tsx#L57-L61
and while changing username `PATCH` is used

---

After reading the implementation of functions used, the problem seem to be in

https://github.com/revoltchat/backend/blob/020f2a1b1ae8cd1f85863ec557d8dd0b3d02a2c6/crates/quark/src/impl/generic/users/user.rs#L157-L173
```rust
    pub async fn update_username(&mut self, db: &Database, username: String) -> Result<()> {
        let username = username.trim().to_string();


        if db.is_username_taken(&username).await? {
            return Err(Error::UsernameTaken);
        }


        self.update(
            db,
            PartialUser {
                username: Some(username),
                ..Default::default()
            },
            vec![],
        )
        .await
    }
```
specifically this line `let username = username.trim().to_string();` where the provided username string is trimmed which results in an empty string

where in account creation the username is inserted directly to the db 
https://github.com/revoltchat/backend/blob/020f2a1b1ae8cd1f85863ec557d8dd0b3d02a2c6/crates/delta/src/routes/onboard/complete.rs#L22-L47
```rust
pub async fn req(
    db: &State<Database>,
    session: Session,
    user: Option<User>,
    data: Json<DataOnboard>,
) -> Result<EmptyResponse> {
    if user.is_some() {
        return Err(Error::AlreadyOnboarded);
    }


    let data = data.into_inner();
    data.validate()
        .map_err(|error| Error::FailedValidation { error })?;


    if db.is_username_taken(&data.username).await? {
        return Err(Error::UsernameTaken);
    }


    let user = User {
        id: session.user_id,
        username: data.username,
        ..Default::default()
    };


    db.insert_user(&user).await.map(|_| EmptyResponse)
}
```

**Comment by I2rys** [comment]
Thank you so much @LedaThemis for clarifying & pinpointing the problem. Let's say the maintainers don't want to remove ```trim()``` as It might cause more problem an alternative solution will be is by checking the username length after It's been trimmed, If It's 0 then we can already already assume that the user want an invisible name so we can just convert the string to invisible characters again.

**Comment by I2rys** [comment]
This has been fixed in the newest version.

**Comment by LedaThemis** [comment]
Partially fixed in https://github.com/revoltchat/backend/commit/80af18761fd3e2727f1d51d66e8a615e17ebdf05 which added more validation to username changing, but the problem is still there in account creation. this is also true for #667 (perhaps this should be a new issue?)

**Comment by I2rys** [comment]
> Partially fixed in [revoltchat/backend@80af187](https://github.com/revoltchat/backend/commit/80af18761fd3e2727f1d51d66e8a615e17ebdf05) which added more validation to username changing, but the problem is still there in account creation. this is also true for #667 (perhaps this should be a new issue?)

I'd say that is a new issue. I will make it. :)

**Comment by I2rys** [comment]
I also managed to bypassed Revolt username validation with \n because insert forgot that you can also use \r but I made a PR to fix it.
